[FEAT] Minimalistic raw maps: Download regions instead of countries

Ebe66 opened the pull request:
## This PR…

- Changes the map selection for downloading and processing from using larger region maps i.e. not downloading the US when only New York is needed to smaller region maps i.e. just download New York.
- closes #260

## Considerations and implementations

Minimalistic map selection uses smaller maps when available resulting in potentially smaller downloads and faster processing (it is faster to split tiles from smaller files) See Issue https://github.com/treee111/wahooMapsCreator/issues/260

Nothing to add except a possibly desired side effect is that in the GUI you can now select regions of a country to create.
Instead of selecting continent Europe and country United-kingdom to create a map of lets say London. You can now select country England and region  Greater-London. I call it a plus others might disagree

## How to test

1. Try python -m wahoomc cli -xy 75/96 from issue https://github.com/treee111/wahooMapsCreator/issues/260 and see that the regions processed has changed to us_connecticut,us_new-jersey,us_new-york from us,us_connecticut,us_new-jersey,us_new-york
2. ...

I can't try the final creation of the maps because I can't get gdal to work. So it still needs some testing.

## Pull Request Checklist
- [ ] Reviewed the [Contributing Guidelines](https://github.com/treee111/wahooMapsCreator/blob/develop/docs/CONTRIBUTING.md)
    + Especially the [Pull-Requests](https://github.com/treee111/wahooMapsCreator/blob/develop/docs/CONTRIBUTING.md#Pull-Requests) section
- [ ] Commits (and commit-messages) are understandable
- [ ] Tested with macOS / Linux
- [x] Tested with Windows

Ebe66 committed:
Minimalistic maps selection and other zoom interval

Minimalistic map selection uses smaller maps when available resulting in potentially smaller downloads and faster processing (it is faster to split tiles from smaller files)

And changed to zoom-interval-conf from 10,0,17 to 12,0,17. This creates smaller sub-files in the resulting maps. This in turn makes sure that for instance Paris can be rendered on the Bolt 1&2 and the roam 1&2

Ebe66 committed:
Removed extra spaces

Ebe66 commented:
@treee111 I installed wmc on another computer and was able to test these modifications. Tested with the x/y coordinates from https://github.com/treee111/wahooMapsCreator/issues/260 (75/96) and with Mittelfranken, both were handled as expected, no download of US or Germany, just the needed states/regions.

treee111 commented:
HI @Ebe66, many, many thanks for the PR!

I stilll wonder how you have been successful in downloading/processing countries/regions with a slash `/` in it:
I had this error due to not being possible to find the geofabrik entry to get the download URL by u`s_connecticut` because only `us/connecticut` is maintained in geofabrik.
I made a PR for this: #276 which might be a preprocessor for your PR.

Have you had the same error? In my opinion this shouldn't be related to OS, i.e. MacOS vs. Windows and happen for you as well.

```
<computer> wahooMapsCreator % python -m wahoomc cli -xy 75/96
INFO:--------------------------------------------------------------------------------
INFO:--------------------------------------------------------------------------------
INFO:# check geofabrik.json file
INFO:# Input X/Y coordinates: 75/96.
INFO:+ Searching for needed maps, this can take a while.
INFO:--------------------------------------------------------------------------------
INFO:# Determine involved/border countries
INFO:+ Involved country: us
INFO:+ Involved country: us_connecticut
INFO:+ Involved country: us_new-jersey
INFO:+ Involved country: us_new-york
INFO:+ Border countries will be processed
INFO:--------------------------------------------------------------------------------
INFO:# check land_polygons.shp file
INFO:--------------------------------------------------------------------------------
INFO:# check countries .osm.pbf files
INFO:+ Checking for old maps and remove them
INFO:+ mapfile for us: up-to-date.
INFO:--------------------------------------------------------------------------------
INFO:# Downloading us_connecticut-latest.osm.pbf file
Traceback (most recent call last):
  File "/opt/anaconda3/envs/gdal-dev/lib/python3.10/runpy.py", line 196, in _run_module_as_main
    return _run_code(code, main_globals, None,
  File "/opt/anaconda3/envs/gdal-dev/lib/python3.10/runpy.py", line 86, in _run_code
    exec(code, run_globals)
  File "<computer>/wahooMapsCreator/wahoomc/__main__.py", line 7, in <module>
    main.run("run")
  File "<computer>/wahooMapsCreator/wahoomc/main.py", line 73, in run
    o_downloader.download_files_if_needed()
  File "<computer>/wahooMapsCreator/wahoomc/downloader.py", line 289, in download_files_if_needed
    self.download_osm_pbf_file()
  File "<computer>/wahooMapsCreator/wahoomc/downloader.py", line 370, in download_osm_pbf_file
    download_file(map_file_path, url)
  File "<computer>/wahooMapsCreator/wahoomc/downloader.py", line 50, in download_file
    if url.split('.')[-1] == 'zip':
AttributeError: 'NoneType' object has no attribute 'split'
```

cnroodbv commented:
What can I say :-) My pull request works but I know that with your version I got the same error. 
For me the solution was i think commenting out these lines, but strangely enough I still have them in my pull request:
        #must_download_maps = [sub.replace(
        #    '/', '_') for sub in must_download_maps]
in geofabrik.py

Proof that the pull request works:  (and look beneath it)
(wmc) C:\Users\ebodd>python -m wahoomc cli -xy 75/96 -fp
INFO:--------------------------------------------------------------------------------
INFO:--------------------------------------------------------------------------------
INFO:# check C:\Users\ebodd\wahooMapsCreatorData\_download\geofabrik.json file
INFO:# Input X/Y coordinates: 75/96.
INFO:+ Searching for needed maps, this can take a while.
INFO:--------------------------------------------------------------------------------
INFO:# Determine involved/border countries
INFO:+ Involved country: us_connecticut
INFO:+ Involved country: us_new-jersey
INFO:+ Involved country: us_new-york
INFO:+ Border countries will be processed
INFO:--------------------------------------------------------------------------------
INFO:# check C:\Users\ebodd\wahooMapsCreatorData\_download\land-polygons-split-4326\land_polygons.shp file
INFO:--------------------------------------------------------------------------------
INFO:# check countries .osm.pbf files
INFO:+ Checking for old maps and remove them
INFO:+ mapfile for us_connecticut: up-to-date.
INFO:+ mapfile for us_new-jersey: up-to-date.
INFO:+ mapfile for us_new-york: up-to-date.
INFO:--------------------------------------------------------------------------------
INFO:# Filter tags from country osm.pbf files
INFO:+ Converting map of us_connecticut to o5m format
INFO:+ Filtering unwanted map objects out of map of us_connecticut
INFO:+ Converting map of us_new-jersey to o5m format
INFO:+ Filtering unwanted map objects out of map of us_new-jersey
INFO:+ Converting map of us_new-york to o5m format
INFO:+ Filtering unwanted map objects out of map of us_new-york
INFO:+ Filter tags from country osm.pbf files: OK, took 108.721 s
INFO:--------------------------------------------------------------------------------
INFO:# Generate land for each coordinate
INFO:+ (tile 1 of 1) Coordinates: 75,96
INFO:+ Generate land for each coordinate: OK, took 14.231 s
INFO:--------------------------------------------------------------------------------
INFO:# Generate sea for each coordinate
INFO:+ (tile 1 of 1) Coordinates: 75,96
INFO:+ Generate sea for each coordinate: OK, took 0.008 s
INFO:--------------------------------------------------------------------------------
INFO:# Split filtered country files to tiles
INFO:+ (tile 1 of 1) Coordinates: 75,96 / us_connecticut
INFO:+ (tile 1 of 1) Coordinates: 75,96 / us_new-jersey
INFO:+ (tile 1 of 1) Coordinates: 75,96 / us_new-york
INFO:+ Split filtered country files to tiles: OK, took 27.591 s
INFO:--------------------------------------------------------------------------------
INFO:# Merge splitted tiles with land, elevation, and sea
INFO:+ (tile 1 of 1) Coordinates: 75,96
INFO:+ Merge splitted tiles with land, elevation, and sea: OK, took 52.763 s
INFO:--------------------------------------------------------------------------------
INFO:# Creating .map files for tiles
INFO:+ (tile 1 of 1) Coordinates: 75,96
INFO:+ Creating .map files for tiles: OK, took 102.684 s
INFO:--------------------------------------------------------------------------------
INFO:# Create: .map.lzma files
INFO:+ Country: 75-96
INFO:+ Copying .map.lzma tiles to output folders
INFO:+ Create .map.lzma files: OK, took 0.040 s
INFO:--------------------------------------------------------------------------------
INFO:# Total time 306.134 s


Your version with the two lines commented out:
Sorry will post it tonight, my internet connection at home is down atm :-(

Humm used my works account I see...
Anyway @treee111 please wait till tonight when I can check. Maybe my PR also doesn't work with American states (instead of coordinates)

treee111 commented:
I tried it out now on Windows, the same error as on my Mac occurs.
Please either run with `-fd` for force download or delete the raw map files from `wahooMapsCreatorData/_download/maps/` to have the tool to download the files again!

The two lines commented out was also my first tryout but as testing a little around it grow to very big changes:
- there are two sections with these two lines in geofabrik
- coding later on use the `/` as directory separator on Mac where other code was not able to find the files then (because they are in `maps/us/` 
- doing it like saving the country and country_w_slash or country and country_w_underscore have also not worked throughout the whole processing

In the end I was able to nail it down to only the one change in the linked PR which takes care of getting the download URL to a geofabrik country/region having a `/` in it.
I think that way we have the least amount of changing to bringing wahooMapsCreator to work with `us/..` stuff and on top of that we have this PR here :-)

treee111 committed:
refactor, delete comments

supported by Copilot ;-)

treee111 committed:
undo changing zoom-interval-conf

- the change is separated from this branch/PR and exists now here: https://github.com/treee111/wahooMapsCreator/commit/792f6c7acd2d9b1aefcc28e1c91150a8bcd43a99

treee111 committed:
update unittests by regions instead of countries

- due to processing smaller parts/ regions instead of whole countries the border-countries/regions, involved countries/regions change

treee111 committed:
update comment to reflect actual coding
